File: src/shady/generator_rewrite.c
#include "generator.h"

static void generate_can_be_default_rewritten_fn(Growy* g, json_object* nodes) {
    shd_growy_append_formatted(g, "static bool can_be_default_rewritten(NodeTag tag) {\n");
    shd_growy_append_formatted(g, "\tswitch (tag) { \n");
    assert(json_object_get_type(nodes) == json_type_array);
    for (size_t i = 0; i < json_object_array_length(nodes); i++) {
        json_object* node = json_object_array_get_idx(nodes, i);
        if (is_recursive_node(node))
            continue;
        String name = json_object_get_string(json_object_object_get(node, "name"));
        shd_growy_append_formatted(g, "\t\tcase %s_TAG: return true;\n", name);
    }
    shd_growy_append_formatted(g, "\t\tdefault: return false;\n");
    shd_growy_append_formatted(g, "\t}\n");
    shd_growy_append_formatted(g, "}\n\n");
}

static void generate_rewriter_default_fns(Growy* g, json_object* nodes) {
    shd_growy_append_formatted(g, "static const Node* recreate_node_identity_generated(Rewriter* rewriter, const Node* node) {\n");
    shd_growy_append_formatted(g, "\tswitch (node->tag) { \n");
    assert(json_object_get_type(nodes) == json_type_array);
    for (size_t i = 0; i < json_object_array_length(nodes); i++) {
        json_object* node = json_object_array_get_idx(nodes, i);

        if (is_recursive_node(node))
            continue;

        String name = json_object_get_string(json_object_object_get(node, "name"));
        String snake_name = json_object_get_string(json_object_object_get(node, "snake_name"));
        void* alloc = NULL;
        if (!snake_name) {
            snake_name = to_snake_case(name);
            alloc = (void*) snake_name;
        }
        shd_growy_append_formatted(g, "\t\tcase %s_TAG: {\n", name);
        json_object* ops = json_object_object_get(node, "ops");
        if (ops) {
            assert(json_object_get_type(ops) == json_type_array);
            shd_growy_append_formatted(g, "\t\t\t%s old_payload = node->payload.%s;\n", name, snake_name);
            shd_growy_append_formatted(g, "\t\t\t%s payload;\n", name);
            shd_growy_append_formatted(g, "\t\t\tmemset(&payload, 0, sizeof(payload));\n");
            for (size_t j = 0; j < json_object_array_length(ops); j++) {
                json_object* op = json_object_array_get_idx(ops, j);
                String op_name = json_object_get_string(json_object_object_get(op, "name"));
                bool list = json_object_get_boolean(json_object_object_get(op, "list"));
                bool ignore = json_object_get_boolean(json_object_object_get(op, "ignore"));
                if (ignore)
                    continue;
                String class = json_object_get_string(json_object_object_get(op, "class"));
                if (!class) {
                    assert(!list);
                    shd_growy_append_formatted(g, "\t\t\tpayload.%s = old_payload.%s;\n", op_name, op_name);
                    continue;
                }
                if (strcmp(class, "string") == 0) {
                    if (list)
                        shd_growy_append_formatted(g, "\t\t\tpayload.%s = shd_strings(rewriter->dst_arena, old_payload.%s.count, old_payload.%s.strings);\n", op_name, op_name, op_name);
                    else
                        shd_growy_append_formatted(g, "\t\t\tpayload.%s = shd_string(rewriter->dst_arena, old_payload.%s);\n", op_name, op_name);
                    continue;
                }

                String class_cap = capitalize(class);
                if (list)
                    shd_growy_append_formatted(g, "\t\t\tpayload.%s = rewrite_ops_helper(rewriter, Nc%s, \"%s\", old_payload.%s);\n", op_name, class_cap, op_name, op_name);
                else
                    shd_growy_append_formatted(g, "\t\t\tpayload.%s = rewrite_op_helper(rewriter, Nc%s, \"%s\", old_payload.%s);\n", op_name, class_cap, op_name, op_name);
                free((void*) class_cap);
            }
            shd_growy_append_formatted(g, "\t\t\treturn %s(rewriter->dst_arena, payload);\n", snake_name);
        } else
            shd_growy_append_formatted(g, "\t\t\treturn %s(rewriter->dst_arena);\n", snake_name);
        shd_growy_append_formatted(g, "\t\t}\n", name);
        if (alloc)
            free(alloc);
    }
    shd_growy_append_formatted(g, "\t\tdefault: assert(false);\n");
    shd_growy_append_formatted(g, "\t}\n");
    shd_growy_append_formatted(g, "}\n\n");
}

void generate(Growy* g, json_object* src) {
    generate_header(g, src);

    json_object* nodes = json_object_object_get(src, "nodes");
    generate_can_be_default_rewritten_fn(g, nodes);
    generate_rewriter_default_fns(g, nodes);
}

File: src/shady/generator_rewrite.py
from src.generator.generator import *
from src.generator.generator_common import *
from src.generator.generator_main import main

def generate_can_be_default_rewritten_fn(g, nodes):
    g.g += "static bool can_be_default_rewritten(NodeTag tag) {\n"
    g.g += "\tswitch (tag) { \n"
    assert type(nodes) is list
    for node in nodes:
        if is_recursive_node(node):
            continue
        name = node["name"]
        g.g += f"\t\tcase {name}_TAG: return true;\n"
    g.g += "\t\tdefault: return false;\n"
    g.g += "\t}\n"
    g.g += "}\n\n"

def generate_rewriter_default_fns(g, nodes):
    g.g += "static const Node* recreate_node_identity_generated(Rewriter* rewriter, const Node* node) {\n"
    g.g += "\tswitch (node->tag) { \n"
    assert type(nodes) is list
    for node in nodes:
        if is_recursive_node(node):
            continue
        name = node["name"]
        snake_name = node["snake_name"]

        g.g += f"\t\tcase {name}_TAG: {{\n"
        ops = node.get("ops")
        if ops is not None:
            assert type(ops) is list
            g.g += f"\t\t\t{name} old_payload = node->payload.{snake_name};\n"
            g.g += f"\t\t\t{name} payload;\n"
            g.g += f"\t\t\tmemset(&payload, 0, sizeof(payload));\n"
            for op in ops:
                op_name = op["name"]
                is_list = op.get("list")
                ignore = op.get("ignore")
                if (ignore):
                    continue
                clazz = op.get("class")
                if clazz is None:
                    assert not is_list
                    g.g += f"\t\t\tpayload.{op_name} = old_payload.{op_name};\n"
                elif clazz == "string":
                    if is_list:
                        g.g += f"\t\t\tpayload.{op_name} = shd_strings(rewriter->dst_arena, old_payload.{op_name}.count, old_payload.{op_name}.strings);\n"
                    else:
                        g.g += f"\t\t\tpayload.{op_name} = shd_string(rewriter->dst_arena, old_payload.{op_name});\n"
                else:
                    class_cap = capitalize(clazz)
                    if is_list:
                        g.g += f"\t\t\tpayload.{op_name} = rewrite_ops_helper(rewriter, Nc{class_cap}, \"{op_name}\", old_payload.{op_name});\n"
                    else:
                        g.g += f"\t\t\tpayload.{op_name} = rewrite_op_helper(rewriter, Nc{class_cap}, \"{op_name}\", old_payload.{op_name});\n"
            g.g += f"\t\t\treturn {snake_name}(rewriter->dst_arena, payload);\n"
        else:
            g.g += f"\t\t\treturn {snake_name}(rewriter->dst_arena);\n"
        g.g += f"\t\t}}\n"
    g.g += "\t\tdefault: assert(false);\n"
    g.g += "\t}\n"
    g.g += "}\n\n"

def generate(g, src):
    generate_header(g, src)

    nodes = src["nodes"]
    generate_can_be_default_rewritten_fn(g, nodes)
    generate_rewriter_default_fns(g, nodes)

main(generate)